feat(infra): enhance server infrastructure with Redis connection pool and observability stack

abretonc7s opened the pull request:
# Description
## Overview
This PR introduces significant infrastructure improvements to our socket server architecture, focusing on enhanced reliability, scalability, and observability. The key changes include:

1. **Redis Connection Management Overhaul**: Replaced singleton Redis client with a properly managed connection pool, improving resource utilization and resilience.
2. **Observability Stack**: Integrated Prometheus, Grafana, Loki, and Promtail for comprehensive monitoring and logging.
3. **Architecture Split**: Separated analytics functionality into a dedicated server to improve modularity and reduce coupling.
4. **Documentation Improvements**: Enhanced setup documentation for both development and scalable deployment scenarios.

## Key Implementation Details

### Redis Improvements
- Implemented Redis connection pooling with configurable minimum (default: 15) and maximum (default: 50) connections
- Added automatic key migration for backward compatibility with legacy key formats
- Enhanced connection retry logic with better error handling
- Implemented health checks with monitoring dashboard integration

### Monitoring Stack
- Added Prometheus for metrics collection from all services on local dev
- Configured Grafana with multiple dashboards for system metrics, Redis health, and application monitoring
- Integrated Loki and Promtail for centralized log collection and visualization
- Added custom metrics for tracking Redis pool usage, key migrations, and handler performance

### Architecture Changes
- Extracted analytics API to a dedicated service (`packages/analytics-server`)
- Added middleware for redirecting analytics traffic to the new dedicated service
- Refactored socket server to support clustered Redis deployments
- Implemented graceful shutdown procedures that properly clean up resources

### Docker Environment Enhancements
- Created two deployment modes:
  1. **Development Mode**: Single instance with auto-reload for development
  2. **Scalable Mode**: Multiple app instances with load balancing for production-like testing
- Added Redis cluster configuration with proper support for horizontal scaling

## Breaking Changes
- Analytics endpoints (`/evt` and `/debug`) are now redirected to a separate service
- Upgraded Redis and Socket.IO dependencies with more robust connection handling
- Changed default analytics server URL to point to the new separate service

## Migration Steps
This PR is designed to maintain backward compatibility, so explicit migration steps shouldn't be necessary. The server will automatically:
1. Detect and migrate legacy Redis keys to the new format when encountered
2. Redirect analytics requests to the new service
3. Handle both old and new client connection patterns

## Testing Instructions
### Development Mode
```bash
# Start background services
docker compose up -d cache prometheus grafana loki promtail

# Run the socket server with auto-reload
docker compose up appdev
```

### Scalable Mode Simulation
```bash
# Initialize Redis Cluster
docker compose up redis-cluster-init

# Start the full stack
docker compose up -d redis-master1 redis-master2 redis-master3 app1 app2 app3 nginx prometheus grafana loki promtail
```

- Access Grafana at http://localhost:3444 (login: gadmin/admin)
- Access Prometheus at http://localhost:9090
- Socket server is available at http://localhost:4000 (dev) or http://localhost:8080 (scalable)

## Benefits & Impact
- Improved reliability through connection pooling and automatic recovery
- Enhanced visibility with comprehensive dashboards for metrics and logs
- Better scalability with proper Redis cluster support
- Cleaner architecture with separation of analytics from core socket functionality
- Improved developer experience with better documentation and development environment

chakra-guy commented on packages/sdk-socket-server-next/src/metrics.ts:
```diff
@@ -385,3 +386,29 @@ export function observeLeaveChannelDuration(duration: number) {
 export function observeCheckRoomDuration(duration: number) {
   checkRoomDuration.observe(duration);
 }
+
+// Add a counter for overall migration progress
+let totalKeysMigrated = 0;
+
+// Add migration metrics to track conversion from old to new Redis key formats
+export const incrementKeyMigration = ({
+  migrationType,
+}: {
+  migrationType: string;
+}) => {
+  totalKeysMigrated++;
+  incrementRedisCacheOperation(`migration-${migrationType}`, true);
+
+  // Log migration progress when reaching certain thresholds
+  if (totalKeysMigrated % 100 === 0) {
+    getLogger().info(`Migration progress: ${totalKeysMigrated} total keys migrated so far`);
+  }
+};
+
+// Add a function to get migration stats for monitoring
+export const getMigrationStats = () => {
+  return {
+    totalKeysMigrated,
+    timestamp: new Date().toISOString(),
+  };
+};
```
all of this should be in prom and grafana

abretonc7s replied:
this is to not loose info, previous migration were ignoring existing keys. Migration is a one time only thing doesnt need to be kept forever and code will be removed after.

chakra-guy replied:
I agree, this is a temp thing, but we could remove the prom metric just as easily as well, right?

I think there are several benefits to use a prom metric here:
- It's accurate as if a crash a redeploy happens, we get to see the correct numbers.
- It works across nodes and gets aggregated
- It can be nicely charted so we can see not just the total number, but the rate

chakra-guy reviewed (changes requested):
please use prom for metrics

chakra-guy reviewed (commented):
A more general note on this PR:

What is the goal of this PR? How do we know that this PR is successful? How can we measure it?

I think we should define these and add metrics for it.

abretonc7s commented:
> A more general note on this PR:
> 
> What is the goal of this PR? How do we know that this PR is successful? How can we measure it?
> 
> I think we should define these and add metrics for it.

Obvisouly fix existing reddit issue or at least give us more insights on the actual root cause... There is a part of monitoring that comes with it. We first fix the problem.

abretonc7s committed:
feat(redis): implement connection pooling and key migration for Redis

- Introduced a connection pool for Redis to improve resource management and reduce connection churn.
- Added backward compatibility for Redis key formats, migrating legacy keys to new formats during access.
- Enhanced metrics for tracking key migration progress and connection pool health.
- Updated Redis client interactions across the application to utilize the new pooling mechanism.
- Implemented health checks and logging for Redis connections to improve operational visibility.

chakra-guy commented:
> > A more general note on this PR:
> > What is the goal of this PR? How do we know that this PR is successful? How can we measure it?
> > I think we should define these and add metrics for it.
> 
> Obvisouly fix existing reddit issue or at least give us more insights on the actual root cause... There is a part of monitoring that comes with it. We first fix the problem.

Totally, I just want to know if the goal is to fix that specific redis problem, how can we measure that it worked?

The metrics that come to my mind that we could track is: "redis command error" or "redis key slot error"

abretonc7s committed:
feat(analytics): add analytics server and client packages

- Introduced new packages for analytics server and client.
- Updated package.json and yarn.lock to include dependencies for the new packages.
- Implemented basic server functionality with event tracking and logging.
- Added ESLint and Prettier configurations for code quality.
- Created README files for both packages.

abretonc7s committed:
feat(protocol): introduce architectural changes for communication layer

- Added a new document outlining the need to split the `sdk-communication-layer` package to resolve dependency conflicts and optimize bundle size for mobile and dApp implementations.
- Implemented a new middleware for analytics redirection in the socket server, enhancing request handling for analytics events.
- Updated the socket server's entry point to include the new analytics middleware.
- Added a clean command for Redis in the socket server's package.json to facilitate easier development and testing.

ecp4224 commented:
> > > A more general note on this PR:
> > > What is the goal of this PR? How do we know that this PR is successful? How can we measure it?
> > > I think we should define these and add metrics for it.
> > 
> > 
> > Obvisouly fix existing reddit issue or at least give us more insights on the actual root cause... There is a part of monitoring that comes with it. We first fix the problem.
> 
> Totally, I just want to know if the goal is to fix that specific redis problem, how can we measure that it worked?
> 
> The metrics that come to my mind that we could track is: "redis command error" or "redis key slot error"

We can track errors using [these logs](https://grafana.o11y.web3factory.consensys.net/explore?schemaVersion=1&panes=%7B%22pfj%22%3A%7B%22datasource%22%3A%22sS52nlJVz%22%2C%22queries%22%3A%5B%7B%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22sS52nlJVz%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bcluster%3D%5C%22mmcx-prd%5C%22%2C+container%3D%7E%5C%22.*metamask.*%5C%22%7D+%7C+detected_level%3D%60error%60%22%2C%22queryType%22%3A%22range%22%2C%22refId%22%3A%22A%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-15m%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=24). This is what I used to verify my previous redis fixes

abretonc7s committed:
refactor(analytics): update server URL references and introduce analytics server URL

- Changed the parameter name from `socketServerUrl` to `analyticsServerUrl` in the `SendAnalytics` function for clarity.
- Added a new constant `DEFAULT_ANALYTICS_SERVER_URL` for the default analytics server URL.
- Updated various components to use `analyticsServerUrl` instead of `communicationServerUrl` for sending analytics events, ensuring consistency across the communication layer.

abretonc7s committed:
refactor(analytics): streamline analytics handling and server URL configuration

- Updated the `initializeConnector` function to include `analyticsServerUrl` for improved clarity.
- Changed the default analytics server URL to match the primary server URL for consistency.
- Removed unused analytics-related code from the socket server, simplifying the analytics API and enhancing maintainability.
- Adjusted the cleanup process to eliminate unnecessary analytics flushing during server shutdown.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into redis-consolidation

abretonc7s committed:
feat(analytics-server): add Dockerfile for analytics server setup

- Introduced a multi-stage Dockerfile for building and running the analytics server.
- Configured the build stage to install dependencies and build the project.
- Set up the runtime stage to install only production dependencies and expose the server on port 2002.
- Updated the default analytics server URL in the configuration to point to localhost for local development.

abretonc7s committed:
feat(redis): implement connection pool and enhance Redis integration

- Replaced singleton Redis client with a managed connection pool, improving connection efficiency.
- Integrated Socket.IO Redis adapter for better cluster support and resolved compatibility issues.
- Added monitoring metrics for Redis pool usage and introduced a health check API endpoint.
- Enhanced stability with graceful shutdown and automatic connection recovery.
- Refactored code to streamline Redis operations and improve maintainability.

abretonc7s committed:
feat(monitoring): integrate Prometheus and Grafana into Docker Compose

- Added Prometheus service for metrics scraping with a pinned version for stability.
- Configured Grafana service for metrics visualization, including default datasource setup.
- Created a new Prometheus configuration file to define scrape targets for app1, app2, and app3.
- Updated README to include setup instructions for the new monitoring stack.

abretonc7s committed:
feat(docker): enhance Docker Compose setup for Redis and monitoring

- Added Redis cluster configuration with multiple master nodes for improved scalability.
- Updated application services to depend on the Redis cluster and configured environment variables for Redis connections.
- Modified Grafana service to expose its UI on a new port and included provisioning for dashboards and data sources.
- Created a default dashboard for Grafana to visualize socket server metrics.
- Revised README to reflect changes in setup instructions and features for both development and scalable environments.

abretonc7s committed:
feat(monitoring): integrate Loki and Promtail for enhanced logging

- Added Loki and Promtail services to the Docker Compose setup for centralized log management.
- Configured Promtail to scrape logs from Docker containers based on labels for better observability.
- Created a new Grafana dashboard for visualizing logs and metrics from the socket server and Redis.
- Updated README with instructions for starting the new logging stack alongside existing services.
- Removed the default Grafana dashboard in favor of a more comprehensive logging dashboard.

abretonc7s committed:
feat: wip

abretonc7s committed:
feat(docs): update README and Grafana dashboards for improved clarity and functionality

- Enhanced the README with clearer instructions for setting up the SDK socket server in development and scalable environments.
- Reformatted bullet points for better readability.
- Updated Grafana dashboard configurations to ensure proper integration with Prometheus and Loki, including adjustments to data source references and layout.
- Improved the handling of log dashboards to ensure consistent data visualization.
- Refactored socket server configuration to utilize a more efficient Redis client setup with expiration handling for channel occupancy.

abretonc7s committed:
fix(docs): update README to include Loki in background services setup

- Revised the setup instructions in the README to specify the inclusion of Loki alongside Redis, Prometheus, and Grafana in the Docker Compose command for starting background services.

abretonc7s committed:
chore: remove migration status endpoint from socket server

- Deleted the migration status endpoint and its associated logic from the socket server codebase, streamlining the server functionality.

abretonc7s committed:
refactor(redis): improve health monitoring and logging logic

- Initialized Redis health status to false to ensure the first success is logged.
- Enhanced logging to provide clearer messages when transitioning between healthy and unhealthy states.
- Consolidated the import of ClientType from a new socket-types module for better organization.
- Updated multiple protocol files to reflect the new import path for ClientType.

abretonc7s committed:
feat: cleanup

abretonc7s committed:
feat: cleanup

ecp4224 commented on packages/sdk-socket-server-next/src/protocol/retrieveMessages.ts:
```diff
@@ -12,10 +13,47 @@ export const retrieveMessages = async ({
+      const legacyMessageData = await pubClient.lrange(legacyQueueKey, 0, -1);
+
+      // If found messages in legacy format, migrate them to new format
+      if (legacyMessageData.length > 0) {
```
what's preventing a migration from happening more than once for a user?

ecp4224 commented on packages/sdk-socket-server-next/src/redis.ts:
```diff
@@ -0,0 +1,492 @@
+ * PooledClientWrapper - A Redis client wrapper that uses the connection pool internally
+ * ... (rest of the class definition as before) ...
+ */
+class PooledClientWrapper {
```
While it does seem easier to work with, this seems odd as a pattern. If any of the protocol messages use `pubClient` more than once, then they may be using different redis clients between different calls (`get`, `set`, etc.), which seems inefficient at best or that it would cause concurrency issues at worse. But since pooled clients don't use auto pipelining, maybe the concurrency issues aren't an actual concern and just the use of multiple clients per call is more of the concern

ecp4224 commented on packages/sdk-socket-server-next/src/socket-types.ts:
```diff
@@ -0,0 +1 @@
+export type ClientType = 'dapp' | 'wallet';
```
Is there a better place for this single type def instead of this file?

ecp4224 commented on packages/sdk-socket-server-next/src/utils.ts:
```diff
@@ -50,31 +50,21 @@ export const setIsShuttingDown = (value: boolean) => {
-  server: Server,
-  analytics: Analytics,
-): Promise<void> => {
+export const cleanupAndExit = async (server: Server): Promise<void> => {
```
should we also clean up and the redis pool?

ecp4224 reviewed (changes requested):
some comments and questions I had while going through the PR

abretonc7s committed:
fix: pr comments

abretonc7s committed:
fix: pr comments

abretonc7s committed:
fix: pipeline

ecp4224 commented on packages/analytics-server/src/index.ts:
```diff
@@ -0,0 +1,171 @@
+// Rate limiting configuration
+const limiter = rateLimit({
+    windowMs: 60 * 1000, // 1 minute
+    max: 100000, // limit each IP to 100,000 requests per windowMs
```
We should still have realistic rate limits even if rate limiting is primarily handled at the infrastructure level. Otherwise, if it's effectively unused then why even have it here?

abretonc7s committed:
feat: pr comments

abretonc7s committed:
feat: lockfile

abretonc7s committed:
fix: pipeline

chakra-guy reviewed (commented):
While I believe these are good high level changes to make and will make the service better, this PR presents significant challenges for review:

**Primary concerns:**
- ~6K lines of changes across multiple disconnected functional areas
- No tests included, making it difficult to verify this works correctly
- The PR description doesn't clearly explain the rationale behind key architectural decisions or sometimes overstates things

### Recommendation: Split into Focused PRs

Breaking this into multiple, focused PRs would significantly improve our ability to review and verify each change:

#### 1. Local Development Environment PR
- Docker environment enhancements for single and multi-instance mode
- Local Prometheus/Grafana/Loki setup
- Scope: Configuration changes only, no business logic modifications
- Purpose: Improve local development experience

#### 2. Analytics Service Extraction PR
- Extract analytics API to a dedicated service (packages/analytics-server)
- Add middleware for redirecting analytics traffic
- Scope: Clean separation of the analytics functionality
- Purpose: Improve modularity and separation of concerns

#### 3. Redis Integration Removal from Analytics Service PR
- Remove Redis dependencies from the extracted analytics service
- Scope: Changes limited to the analytics service implementation
- Purpose: Further decoupling of components

#### 4. Redis Connection Pooling PR
- Replace singleton Redis client with properly managed connection pool
- Add connection retry logic and error handling
- Scope: Redis connection management only
- Purpose: Improve resilience and resource utilization

#### 5. Redis Key Migration PR
- Implement automatic key migration for backward compatibility
- Add proper Prometheus metrics for monitoring the migration
- Scope: Key format changes and migration logic only
- Purpose: Support legacy key formats while moving to new architecture

### Specific Implementation Concerns

1. **Monitoring Claims**: The PR description suggests implementation of Prometheus, Grafana, and Loki as new additions, but these monitoring tools were already set up in production/dev environments months ago. The actual contribution appears limited to configuring a local development setup.

2. **Redis Key Migration Metrics**: The implementation uses a global variable counter exposed via an API endpoint rather than Prometheus metrics, which is inconsistent with our existing observability stack. As already noted in earlier comments, we should be using Prometheus for all metrics.

3. **Redis Key Migration**: Could you please also explain in the PR what is the problem with the current key and why would this new format fix the issue?

Thanks!

abretonc7s commented:
> While I believe these are good high level changes to make and will make the service better, this PR presents significant challenges for review:
> 
> **Primary concerns:**
> 
> * ~6K lines of changes across multiple disconnected functional areas
> * No tests included, making it difficult to verify this works correctly
> * The PR description doesn't clearly explain the rationale behind key architectural decisions or sometimes overstates things
> 
> ### Recommendation: Split into Focused PRs
> Breaking this into multiple, focused PRs would significantly improve our ability to review and verify each change:
> 
> #### 1. Local Development Environment PR
> * Docker environment enhancements for single and multi-instance mode
> * Local Prometheus/Grafana/Loki setup
> * Scope: Configuration changes only, no business logic modifications
> * Purpose: Improve local development experience
> 
> #### 2. Analytics Service Extraction PR
> * Extract analytics API to a dedicated service (packages/analytics-server)
> * Add middleware for redirecting analytics traffic
> * Scope: Clean separation of the analytics functionality
> * Purpose: Improve modularity and separation of concerns
> 
> #### 3. Redis Integration Removal from Analytics Service PR
> * Remove Redis dependencies from the extracted analytics service
> * Scope: Changes limited to the analytics service implementation
> * Purpose: Further decoupling of components
> 
> #### 4. Redis Connection Pooling PR
> * Replace singleton Redis client with properly managed connection pool
> * Add connection retry logic and error handling
> * Scope: Redis connection management only
> * Purpose: Improve resilience and resource utilization
> 
> #### 5. Redis Key Migration PR
> * Implement automatic key migration for backward compatibility
> * Add proper Prometheus metrics for monitoring the migration
> * Scope: Key format changes and migration logic only
> * Purpose: Support legacy key formats while moving to new architecture
> 
> ### Specific Implementation Concerns
> 1. **Monitoring Claims**: The PR description suggests implementation of Prometheus, Grafana, and Loki as new additions, but these monitoring tools were already set up in production/dev environments months ago. The actual contribution appears limited to configuring a local development setup.
> 2. **Redis Key Migration Metrics**: The implementation uses a global variable counter exposed via an API endpoint rather than Prometheus metrics, which is inconsistent with our existing observability stack. As already noted in earlier comments, we should be using Prometheus for all metrics.
> 3. **Redis Key Migration**: Could you please also explain in the PR what is the problem with the current key and why would this new format fix the issue?
> 
> Thanks!

the redis key migration is to force each key to stay in the same host, seems to be a redis specific format.

abretonc7s committed:
fix: pipeline
